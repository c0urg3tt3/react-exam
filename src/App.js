import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { connect } from 'react-redux'

import './App.css'

import fetchJedi from './redux/actions/jedi/fetch'

import AppHeader from './components/app-header/component'
import AppLogo from './components/app-logo/component'
import AppTitle from './components/app-title/component'
import JediListItem from './components/jedi-list-item/component'
import JediListEmpty from './components/jedi-list-empty/component'

export class App extends Component {
  componentWillMount() {
    this.props.fetchJedi()
  }

  render() {
    const { jedies } = this.props

    return (
      <div className="app">
        <AppHeader>
          <AppLogo/>
          <AppTitle/>
        </AppHeader>
        {(!jedies.length && (
          <JediListEmpty>
          {({message, quote, autor}) => (
            <div className="jedi-quote">
              <p>{message}</p>
              <blockquote>
                <p>{quote}</p>
                <footer>{autor}</footer>
              </blockquote>
            </div>
          )}
          </JediListEmpty>
        ))
        || (
          <ul className="app-list">
          {
            jedies.map((jedi, index) => (
              <JediListItem key={index} jedi={jedi}/>
            ))
          }
          </ul>
        )}
      </div>
    )
  }
}

App.propTypes = {
  jedies: PropTypes.array,
  isFetching: PropTypes.bool,
  isFetchingError: PropTypes.bool,
  fetchJedi: PropTypes.func,
}

App.defaultProps = {
  jedies: [],
  isFetching: false,
  isFetchingError: true,
  fetchJedi: noop => noop
}

function mapStateToProps(state) {
  return { ...state.jedi.list }
}

function mapDispatchToProps(dispatch) {
  return {
    fetchJedi: () => dispatch(fetchJedi()),
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(App)
