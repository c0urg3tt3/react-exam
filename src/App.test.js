import React from 'react'
import { shallow } from 'enzyme'

import { App } from './App'

describe('App', () => {

  it('should render component', () => {
    const wrapper = shallow(<App/>)
    console.log(wrapper.debug())
    expect(wrapper.exists()).toEqual(true)
    expect(wrapper.is('div')).toEqual(true)
    expect(wrapper.hasClass('App')).toEqual(true)

    const header = wrapper.find('AppHeader')
    expect(header.exists()).toEqual(true)

      const img = header.find('AppLogo')
      expect(img.exists()).toEqual(true)
      expect(img.prop('imageURL')).toEqual("logo_jedi_order.svg")
      expect(img.prop('isAnimate')).toEqual(true)

      const h1 = wrapper.find('AppTitle')
      expect(h1.exists()).toEqual(true)
      expect(h1.prop('title')).toEqual("Jedi List")

    const list = wrapper.find('.App-list')
    expect(list.exists()).toEqual(true)
    expect(list.is('ul')).toEqual(true)

    const listItems = wrapper.find('.App-list-item')
    expect(listItems.length).toEqual(0)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should call fetchJedi when commponent mount', () => {
    const appProps = {
      fetchJedi: jest.fn()
    }
    const wrapper = shallow(<App {...appProps}/>)

    expect(appProps.fetchJedi.mock.calls.length).toBe(1)
  })

  it('should render list item when jedi fetch success', () => {
    const appProps = {
      jedi: [
        {id: 42, name: 'Jar Jar Bings'},
        {id: 1337, name: 'Anakin Skywalkers'}
      ]
    }
    const wrapper = shallow(<App {...appProps}/>)

    const listItems = wrapper.find('.App-list-item')
    expect(listItems.length).toEqual(2)
    expect(listItems.first().is('li')).toEqual(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
