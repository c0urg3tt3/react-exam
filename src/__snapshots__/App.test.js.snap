// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render component 1`] = `"<div class=\\"App\\"><header class=\\"App-header\\"><img src=\\"logo.svg\\" class=\\"App-logo\\" alt=\\"logo\\"/><h1 class=\\"app-title\\">Jedi List</h1></header><ul class=\\"App-list\\"></ul></div>"`;

exports[`App should render list item when jedi fetch success 1`] = `"<div class=\\"App\\"><header class=\\"App-header\\"><img src=\\"logo.svg\\" class=\\"App-logo\\" alt=\\"logo\\"/><h1 class=\\"app-title\\">Jedi List</h1></header><ul class=\\"App-list\\"><li class=\\"App-list-item\\">Jedi: id: 42 name: Jar Jar Bings</li><li class=\\"App-list-item\\">Jedi: id: 1337 name: Anakin Skywalkers</li></ul></div>"`;
